<a href="https://colab.research.google.com/github/leehyell/AIStudy24/blob/master/%EC%95%99%EC%83%81%EB%B8%94.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###앙상블
* 단어 그대로 **'여러 단순한 모델을 결합하여 정확한 모델을 만드는 방법'**이다.

* 단어와 그 의미
    * 정형 데이터: 지금까지 학습한 수치 자료가 있는 값.
    * 비정형 데이터: 데이터 베이스나 엑셀로 표현하기 어려운 데이터.
        * 텍스트 데이터, 디카 사진, mp3 등 → 신경망 알고리즘(딥러닝)
    * 랜덤 포레스트: 결정 트리를 랜덤하게 만들어 결정트리(나무) 숲을 만듦.
        * → 최종 예측
<br/><br/>
    * 1000개의 샘플이 들어있는 가방에서 100개 샘플을 뽑을 때, 1개를 뽑고 뽑았던 1개를 다시 가방에 넣음.
    * 중복된 샘플을 뽑을 수 있음. → 부트스트랩 샘플이라고 함.
    * 부트스트랩: 데이터 세트에서 중복을 허용하여 데이터를 샘플링.

* 분류 모델인 RandomForestClassifier는
    * 기본적으로 전체 특성 갯수의 제곱근(선형회귀;곡선형) 만큼 특성을 선택.
    * 즉, 4개의 특성이 있다면 노드마다 2개를 랜덤하게 선택해 사용.
    * 다만 회귀 모델인 RandomForestRegressor는 전체 특성을 사용.
<br/><br/>
    * 사이킷 런의 랜덤 포레스트는 기본적으로 100개의 결정 트리를 이런 방식으로 훈련한다.
    * 그 다음 분류일 때는 각 트리의 클래스별 확률을 평균하여 가장 높은 확률을 가진 클래스를 예측으로 삼는다.
    * 회귀일 때는 단순히 각 트리의 예측을 평균함.
    * 분류: 샘플을 몇 개의 클래스 중 하나로 분류하는 문제.
    * 회귀: 임의의 어떤 숫자를 예측하는 문제.

In [71]:
!pip install xgboost
!pip install pycaret

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 169.6/169.6 kB 11.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.1/486.1 kB 25.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.8/106.8 kB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.8/21.8 MB 67.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.0/82.0 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.2/302.2 kB 21.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 81.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 84.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.7/80.7 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 51.6 MB/s eta 0:00:00
   ━━━

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

wine = pd.read_csv('https://bit.ly/wine_csv_data')

data = wine[['alcohol','sugar','pH']].to_numpy()
target = wine['class'].to_numpy()
print(data)
print('-'*23)
print(target)
#0이면 레드, 1이면 화이트 와인

[[ 9.4   1.9   3.51]
 [ 9.8   2.6   3.2 ]
 [ 9.8   2.3   3.26]
 ...
 [ 9.4   1.2   2.99]
 [12.8   1.1   3.34]
 [11.8   0.8   3.26]]
-----------------------
[0. 0. 0. ... 1. 1. 1.]


In [2]:
#훈련용, 테스트용 나누기
#훈련용 80% / 테스트용 20%
train_input, test_input, train_target, test_target = train_test_split(data, target, test_size=0.2, random_state=42)

print(train_input.shape, test_input.shape)
#원래 5197개였던 훈련세트가 4,157개로 줄고, 검증 세트가 1,040개가 됨.

(5197, 3) (1300, 3)


In [3]:
#크로스 검증용
from sklearn.model_selection import cross_validate
#100개의 결정 트리 사용
from sklearn.ensemble import RandomForestClassifier

#n_jobs=-1 → 모든 cpu 사용
rf = RandomForestClassifier(n_jobs=-1, random_state=42)
#return_train_score=True : 검증 점수와 훈련 세트에 대한 점수도 리턴
scores = cross_validate(rf, train_input, train_target, return_train_score=True, n_jobs=-1)

#train_score 과대적합; 0.9973541965122431
print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9973541965122431 0.8905151032797809


In [4]:
#RandomForest는 결정 트리의 앙상블이기 때문에 결정 트리가 제공하는 중요한 매개 변수를 모두 제공한다.
#결정 트리의 큰 장점 중 하나인 특성 중요도를 계산한다.
#랜덤 포레스트의 특성 중요도는 각 결정 트리의 특성 중요도를 취합한 것이다.
#훈련 후에 중요도를 출력해보자.
rf.fit(train_input, train_target)
print(rf.feature_importances_)

[0.23167441 0.50039841 0.26792718]


###RandomForest 중요도 결과
* [0.23167441 0.50039841 0.26792718] 랜덤포레스트        - 현재 결과
* [0.12345626 0.86862934 0.0079144 ] LogisticRegression  - 이전 결과
*   alchol      sugar       pH
* 당도의 중요도가 감소하고, 알콜 도수와 pH 중요도가 상승함.
    * 이유는 랜덤 포레스트가 특성 일부를 랜덤하게 선택해 결정 트리를 훈련함(중복 허용)
* 하나의 특성에 과도하게 집중하지 않고 좀 더 많은 특성이 훈련에 기여할 기회가 됨.
    *└ 과대 적합을 줄이게 됨.

In [5]:
#RandomForestClassifier 기능 중에 자체적으로 모델을 평가하는 점수를 얻을 수 있다.
#OOB(Out Of Bag) 부트스트랩에 포함되지 않고 남은 샘플: 결정 트리 평가용.(검증 세트로 활용)
#oob_score=True → 랜덤 포레스트는 각 결정 트리의 OOB 점수를 평균하여 출력.(oob_score_)
rf = RandomForestClassifier(oob_score=True, n_jobs=-1, random_state=42)
rf.fit(train_input, train_target)
#교차 검증 점수와 비슷한 결과를 얻음.
print(rf.oob_score_)
#OOB 점수를 사용하면 교차 검증을 대신할 수 있어 결과적으로 훈련 세트에 더 많은 샘플을 사용할 수 있음.

0.8934000384837406


###엑스트라 트리(extra tree)
* 100개의 결정 트리를 훈련
    * 하지만 부트스트랩 샘플을 사용하지 않는다.(전체 훈련세트 사용)
* 대신 노드를 분할할 때 가장 좋은 분할을 찾는 것이 아닌 무작위로 분할
* 엑스트라 트리의 결정 트리 splitter='random'

In [6]:
#ExtraTreesClassifier 엑스트라 트리
from sklearn.ensemble import ExtraTreesClassifier

et = ExtraTreesClassifier(n_jobs=-1, random_state=42)
scores = cross_validate(et, train_input, train_target, return_train_score=True, n_jobs=-1)

#RandomForestClassifier와 결과가 비슷함. 대신 속도가 빠름.
#RandomForestClassifier: 6초, ExtraTreesClassifier 2초
print(np.mean(scores['train_score']), np.mean(scores['test_score']))

0.9974503966084433 0.8887848893166506


In [7]:
et.fit(train_input, train_target)
print(et.feature_importances_)

[0.20183568 0.52242907 0.27573525]


### 엑스트라 트리 중요도
* [0.20183568 0.52242907 0.27573525] 엑스트라 트리       - 현재 결과
* [0.23167441 0.50039841 0.26792718] 랜덤포레스트        - 이전 결과
* [0.12345626 0.86862934 0.0079144 ] LogisticRegression  - 이전 결과
* 특성 중요도를 보면 결정 트리보다 당도에 대한 의존성이 작다.

###그레디언트 부스팅(Gradient Boosting); 기울기
* 깊이가 얕은 결정 트리를 사용하여 이전 트리의 오차를 보완하는 방식으로 앙상블
* 사이킷 런의 GradientBoostingClassifier는 기본적으로 깊이가 3인 결정 트리를 100개 사용
* 때문에 과대적합에 강하고 일반적인 높은 일반화 성능을 기대함.
* Gradient는 **경사 하강법**을 사용하여 트리를 앙상블에 추가함.
    * 경사하강법: 가파른 경사를 빠르게 내려오는 방법.
* 분류: 로지스틱 손실 함수, 회귀에서는 평균 제급 오차 함수를 사용.
* 경사 하강법 손실 함수를 산으로 정의하고 가장 낮은 곳으로 찾아 내려오는 과정.
* 가장 낮은 곳으로 내려오는 방법은 모델의 가중치와 절편을 조금씩 바꾸는 것.
* 결정 트리를 계속 추가하면서 가장 낮은 곳을 찾아 이동, 손실 함수의 낮은 곳으로 천천히 이동.

In [8]:
#GradientBoostingClassifier 그레디언트 부스팅
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(random_state=42)
scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)
print(np.mean(scores['train_score']), np.mean(scores['test_score']))
#과대적합 해결(0.8881086892152563 0.8720430147331015)

0.8881086892152563 0.8720430147331015


In [9]:
#n_estimators=500   : 결정 트리의 갯수를 500개로 늘림(기본 100)
#learning_rate=0.2  : 학습률 기본값 0.1
gb = GradientBoostingClassifier(n_estimators=500, learning_rate=0.2, random_state=42)

scores = cross_validate(gb, train_input, train_target, return_train_score=True, n_jobs=-1)

#교차 검증 점수 확인
print(np.mean(scores['train_score']), np.mean(scores['test_score']))
#94.64%, 87.80%
#점수 차이가 10% 이상 나지 않으면 괜찮음(과대적합 X)

0.9464595437171814 0.8780082549788999


In [10]:
gb.fit(train_input, train_target)
print(gb.feature_importances_)

#[0.15887763 0.6799705  0.16115187]
#    알콜       당도         pH
#당도에 대한 의존도가 낮음.
#그레이디언트 부스팅이 랜덤 포레스트보다 조금 더 높은 성능이 있다.
#하지만 순서대로 트리를 추가하기 때문에 훈련 속도는 조금 느리다.
#GradientBoostingClassifier에는 n_jobs=-1 매개값이 없다.(구형 모델이라.)

[0.15887763 0.6799705  0.16115187]


### 그레디언트 부스팅 중요도
* [0.15887763 0.6799705  0.16115187] 그레디언트 부스팅   - 현재 결과
* [0.20183568 0.52242907 0.27573525] 엑스트라 트리       - 이전 결과
* [0.23167441 0.50039841 0.26792718] 랜덤포레스트        - 이전 결과
* [0.12345626 0.86862934 0.0079144 ] LogisticRegression  - 이전 결과
* 특성 중요도를 보면 결정 트리보다 당도에 대한 의존성이 작다.

###히스토그램 기반
* 히스토그램 기반 그레디언트 부스팅(Histogram Gradient): 그레이디언트 부스팅의 개선 버전
* 입력 특성을 256 구간으로 나눔
    * 노드를 분할할 때 최적의 분할을 매우 빠르게 찾을 수 있다.
* 특히 256 구간 중에서 하나를 떼어놓고 누락된 값을 위해서 사용함.
* HistGradientBoostingClassifier는 기본 매개 변수에서 안정적인 성능을 얻을 수 있다.
* HistGradientBoostingClassifier에는 트리의 개수를 지정하는데 n_estimators 대신 max_iter를 사용함
    * 성능 향상용

In [11]:
#from sklearn.ensemble import enable_hist_gradient_boosting 아직 개발 중.
from sklearn.ensemble import HistGradientBoostingClassifier#Classifier: 분류용

hgb = HistGradientBoostingClassifier(random_state=42)
scores = cross_validate(hgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))
#과대 적합 억제 성공(93.21%, 88.01%)

0.9321723946453317 0.8801241948619236


In [12]:
from sklearn.inspection import permutation_importance
#permutation_importance: 특성 중요도 확인
#└ 특성을 하나씩 랜덤하게 섞어서 모델의 성능이 변화되는지 관찰


#히스토그램 기반의 훈련을 진행 → 훈련세트에서 특성 중요도를 계산해보자.
hgb.fit(train_input, train_target)

#n_repeats=10: 랜덤하게 섞을 수 있는 횟수(기본 5)
result = permutation_importance(hgb, train_input, train_target, n_repeats=10, random_state=42, n_jobs=-1)

print(result.importances_mean)
#훈련 세트에서의 특성 중요도
#   중요도      평균     표준편차
#[0.08876275 0.23438522 0.08027708]

[0.08876275 0.23438522 0.08027708]


In [13]:
#이번에는 테스트 세트에 결과를 확인해보자.
result = permutation_importance(hgb, test_input, test_target, n_repeats=10, random_state=42, n_jobs=-1)

print(result.importances_mean)
#테스트 세트에서의 특성 중요도
#   중요도      평균     표준편차
#[0.05969231 0.20238462 0.049     ]

[0.05969231 0.20238462 0.049     ]


### 히스토그램 기반 그레이언트 부스팅 중요도
* [0.05969231 0.20238462 0.049     ] 히스토그램 테스트   - 현재 결과
* [0.08876275 0.23438522 0.08027708] 히스토그램 훈련용   - 현재 결과
* [0.15887763 0.6799705  0.16115187] 그레디언트 부스팅   - 이전 결과
* [0.20183568 0.52242907 0.27573525] 엑스트라 트리       - 이전 결과
* [0.23167441 0.50039841 0.26792718] 랜덤포레스트        - 이전 결과
* [0.12345626 0.86862934 0.0079144 ] LogisticRegression  - 이전 결과

In [15]:
#HistGradientBoostingGegressor 히스토그램 기반 그레이디언트 부스팅의 회귀 버전
#사이킷 말고도 그레이디언트 부스팅 알고리즘을 구현한 라이브러리가 다수 존재
#XGBoost 대표적임.

from xgboost import XGBClassifier

#tree_method='hist' 히스토그램 기반 그레이디언트 부스팅용
xgb = XGBClassifier(tree_method='hist', random_state=42)
scores = cross_validate(xgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))
#95.58% 87.82%

0.9558403027491312 0.8782000074035686


In [17]:
#LGBMClassifier ms에서 만든 LightGBM^2
from lightgbm import LGBMClassifier

lgb = LGBMClassifier(random_state=42)
scores = cross_validate(lgb, train_input, train_target, return_train_score=True, n_jobs=-1)

print(np.mean(scores['train_score']), np.mean(scores['test_score']))
#93.58% 88.01%

0.935828414851749 0.8801251203079884
